### PR TITLE
CLOUDSTACK-8628: kvm: Disable Fencing when no NFS storage pools are p…

### PR DESCRIPTION
**Pull request by wido**
…resent

On NFS we write a heartbeat, but without those we can not safely
fence off a host.

If we fence without knowing about a heartbeat we can cause a split-brain situation.

Signed-off-by: Wido den Hollander <wido@widodh.nl>

### CODE REVIEW

**Comment by wilderrodrigues** [comment]
Hi @wido 

Could you please add an extra test to cover the "pools.size() == 0)"?

Cheers,
Wilder

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtFenceCommandWrapper.java [hunk 1]
```diff
@@ -48,6 +48,18 @@ public Answer execute(final FenceCommand command, final LibvirtComputingResource
...
 
         final List<NfsStoragePool> pools = monitor.getStoragePools();
+
+        /**
+         * We can only safely fence off hosts when we use NFS
+         * On NFS primary storage pools hosts continuesly write
+         * a heartbeat. Disable Fencing Off for hosts without NFS
+         */
+        if (pools.size() == 0) {
+            String logline = "No NFS storage pools found. No way to safely fence " + command.getVmName() + " on host " + command.getHostGuid();
```
Comment (bhaisaab) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtFenceCommandWrapper.java]: maybe add an alert for admin here, _alertMgr.sendAlert?

**Comment by wilderrodrigues** [comment]
@wido any news on this PR?

**Comment by wido** [comment]
Working on a additional commit which brings in more logging and a unit test.

**Comment by wilderrodrigues** [comment]
Travis failed due to NPE:

Tests in error: 
  KVMFencerTest.testWithFailingFence:135 » NullPointer
  KVMFencerTest.testWithSingleHost:75 » NullPointer
  KVMFencerTest.testWithSingleHostDown:88 » NullPointer
  KVMFencerTest.testWithTimeoutingFence:158 » NullPointer

**Comment by wido** [comment]
Indeed, I noticed that to late. Fixing the tests.

**Comment by wido** [comment]
The tests have been fixed, Mocking issue in the test.

**Comment by DaanHoogland** [comment]
LGTM, how about the alert @bhaisaab asked for?

**Comment by wido** [comment]
@DaanHoogland It is in the KVMFencer in the MGMT server. The Agent can't send an alert, so I had to do this in the mgmt server.

**Comment by DaanHoogland** [comment]
@wido thanks for the clarification. as said LGTM

**Pull request closed without merging**